python/ray/autoscaler/node_provider.py

```diff
@@ -276,3 +278,92 @@ def get_command_runner(self,
             return DockerCommandRunner(docker_config, **common_args)
         else:
             return SSHCommandRunner(**common_args)
+
+class NodeProviderCache:
+    """Base class for storing cached node information, used to avoid excessive
+    API calls to cloud providers in some cases.
+
+    This implementation stores all information in memory. 
+    """
+
+    def __init__(self):
+        # tag_map is supposed to be the superset of node_map.
```

Comment:
What do you mean by this?

Reply:
Unfortunately tags and nodes are not always updated together.
I think there are cases there would be tag informations about a node in tag_map, but no corresponding entry in node_map. Let me update this comment